File: iOSNetworkExperiment/NetworkExperiment/WikipediaFetcher.swift
//
//  WikipediaFetcher.swift
//  NetworkExperiment
//

import Alamofire
import SwiftyJSON
import Fetcher

class WikipediaFetcher {
    private static let url = "https://en.wikipedia.org/w/api.php"
    private let fetcher = Fetcher()

    private let onResponse: () -> ()
    private let showResult: (String) -> ()

    init (onResponse: @escaping () -> (),
          showResult: @escaping (String) -> ()) {
        self.onResponse = onResponse
        self.showResult = showResult
    }

    public func fetchData(searchText: String)
        -> DataRequest {
            
            let parameters: Parameters = [
                "action": "query",
                "format": "json",
                "list": "search",
                "srsearch": searchText
            ]

            let request = fetcher.fetch(url: WikipediaFetcher.url,
                                        parameters: parameters,
                                        onResponse: self.onResponse,
                                        onSuccess: self.onSuccess,
                                        onFailure: self.onError)

            return request
    }

    private func onSuccess(data: Any) {
        let swiftyJsonVar = JSON(data)
        let countObj = swiftyJsonVar["query"]["searchinfo"]["totalhits"]
        guard let count = countObj.int else {
            self.showResult("No data found")
            return
        }
        self.showResult("Count is \(count)")

    }

    private func onError(message: String) {
        self.showResult(message)
    }
}

File: AndroidNetworkExperiment/wikipediafetcher/src/main/java/com/elyeproj/wikipediafetcher/WikipediaFetcher.kt
package com.elyeproj.wikipediafetcher

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.HttpUrl
import com.elyeproj.fetcher.Fetcher
import okhttp3.Interceptor

class WikipediaFetcher(userAgentInterceptor: Interceptor) {

    object Model {
        data class Result(val query: Query)
        data class Query(val searchinfo: SearchInfo)
        data class SearchInfo(val totalhits: Int)
    }

    private val fetcher by lazy {
        Fetcher(userAgentInterceptor)
    }

    private val httpUrlBuilder by lazy {
        HttpUrl.Builder()
            .scheme("https")
            .host("en.wikipedia.org")
            .addPathSegment("w")
            .addPathSegment("api.php")
            .addQueryParameter("action", "query")
            .addQueryParameter("format", "json")
            .addQueryParameter("list", "search")
    }

    fun performFetch(
        searchText: String,
        onPerformFetch: () -> Unit,
        onSuccess: (String) -> Unit,
        onFailure: (String) -> Unit
    ): Disposable {
        onPerformFetch()
        val httpUrl = httpUrlBuilder.addQueryParameter("srsearch", searchText).build()

        return Single.just(searchText)
            .map { fetcher.fetchOnBackground(httpUrl, ::onSuccess) }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { onSuccess("Count is $it") },
                { onFailure(it.localizedMessage) })
    }

    private fun onSuccess(raw: String?): String {
        return try {
            val result = Gson().fromJson(raw, Model.Result::class.java)
            result.query.searchinfo.totalhits.toString()
        } catch (exception: JsonSyntaxException) {
            "No data found "
        }
    }
}